{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ce7fcac0-af40-485a-b091-6103f523bcab",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d54d558a-9061-415f-8fe3-c77fcee5f243",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting openpyxl\n",
      "  Downloading openpyxl-3.1.5-py2.py3-none-any.whl.metadata (2.5 kB)\n",
      "Collecting et-xmlfile (from openpyxl)\n",
      "  Downloading et_xmlfile-2.0.0-py3-none-any.whl.metadata (2.7 kB)\n",
      "Downloading openpyxl-3.1.5-py2.py3-none-any.whl (250 kB)\n",
      "Downloading et_xmlfile-2.0.0-py3-none-any.whl (18 kB)\n",
      "Installing collected packages: et-xmlfile, openpyxl\n",
      "Successfully installed et-xmlfile-2.0.0 openpyxl-3.1.5\n"
     ]
    }
   ],
   "source": [
    "!pip install openpyxl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a0c19bc8-7d1c-4734-80c7-c6c8a207fa34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# statistical tests TET2\n",
    "\n",
    "df = pd.read_excel(r\"D:\\Users\\albao\\Downloads\\TET2_predictions_annotationsall_char.xlsx\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bb94c216-51c2-47bc-9d7a-972cf8bec52f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Unnamed: 0',\n",
       " 'SAMPLE',\n",
       " 'FILTER',\n",
       " 'VAR_ID',\n",
       " 'CHR',\n",
       " 'POS',\n",
       " 'REF',\n",
       " 'ALT',\n",
       " 'GENE',\n",
       " 'ALLELE',\n",
       " 'VARIANT_CLASS',\n",
       " 'BIOTYPE',\n",
       " 'TYPE',\n",
       " 'FEATURE',\n",
       " 'EXON',\n",
       " 'HGVSc',\n",
       " 'HGVSp',\n",
       " 'cDNA_position',\n",
       " 'CDS_position',\n",
       " 'Protein_position',\n",
       " 'Amino_acids',\n",
       " 'Codons',\n",
       " 'STRAND',\n",
       " 'REFSEQ_MATCH',\n",
       " 'GIVEN_REF',\n",
       " 'USED_REF',\n",
       " 'BAM_EDIT',\n",
       " 'IMPACT',\n",
       " 'CONSEQUENCE',\n",
       " 'SOMATIC',\n",
       " 'PHENO',\n",
       " 'CLIN_SIG',\n",
       " 'COSMIC_MATCH',\n",
       " 'HEURISTIC',\n",
       " 'PREVIOUSLY_IDENTIFIED',\n",
       " 'WHITELIST',\n",
       " 'ARTIFACT',\n",
       " 'SIFT_DESC',\n",
       " 'SIFT_SCORE',\n",
       " 'POLYPHEN_DESC',\n",
       " 'POLYPHEN_SCORE',\n",
       " 'CADD_PHRED',\n",
       " 'CADD_RAW',\n",
       " 'REF_DEPTH_VIS1',\n",
       " 'ALT_DEPTH_VIS1',\n",
       " 'F1R2_VIS1',\n",
       " 'F2R1_VIS1',\n",
       " 'REF_DEPTH_VIS3',\n",
       " 'ALT_DEPTH_VIS3',\n",
       " 'F1R2_VIS3',\n",
       " 'F2R1_VIS3',\n",
       " 'VAF_VIS1',\n",
       " 'VAF_VIS3',\n",
       " 'VAF_ABSOLUTA',\n",
       " 'VAF_RELATIVA',\n",
       " 'VAF_RATIO',\n",
       " 'CHIP_driver',\n",
       " 'FU_yrs',\n",
       " 'AER',\n",
       " 'log_positive_AER',\n",
       " 'log_AER',\n",
       " 'N',\n",
       " 'doubling_time',\n",
       " 'doubling_time_wo_outliers',\n",
       " 'log_ratio_VAF_years',\n",
       " 'VAF_absolute_yrs',\n",
       " 'growth_lm_VAF',\n",
       " 'growth_lm_log_VAF',\n",
       " 'fitness_exp',\n",
       " 'region_normal',\n",
       " 'region_alt',\n",
       " 'POS.1',\n",
       " 'exon',\n",
       " 'Amino_acids.1',\n",
       " 'IMPACT.1',\n",
       " 'VARIANT_CLASS.1',\n",
       " 'CpG variant',\n",
       " 'PHYLO',\n",
       " 'PHASTCON',\n",
       " 'cancer score',\n",
       " 'cancer score transcript',\n",
       " 'cys-rich domain',\n",
       " 'DSBH',\n",
       " 'disordered',\n",
       " 'phosphoserine',\n",
       " 'phosphothreonine',\n",
       " 'polar residues',\n",
       " 'lys-gly crosslink',\n",
       " 'Assymetric dimethylargine',\n",
       " 'chain',\n",
       " 'interaction with DNA',\n",
       " 'SPI1_score',\n",
       " 'RAD21_score',\n",
       " 'GATA2_score',\n",
       " 'GATA3_score',\n",
       " 'YY1_score',\n",
       " 'EP300_score',\n",
       " 'BATF_score',\n",
       " 'CTCF_score',\n",
       " 'EED_score',\n",
       " 'JUNB_score',\n",
       " 'POLR2A_score',\n",
       " 'POU2F2_score',\n",
       " 'IRF4_score',\n",
       " 'RCOR1_score',\n",
       " 'SMARCA4_score',\n",
       " 'Dnase cluster',\n",
       " 'GC%',\n",
       " 'CPG',\n",
       " 'UKB',\n",
       " 'H3K4Me1',\n",
       " 'H3K4Me3',\n",
       " 'H3K27ac',\n",
       " 'cCREs score',\n",
       " 'cCREs label',\n",
       " 'geneHancer score',\n",
       " 'geneHancer elementType',\n",
       " 'swScore',\n",
       " 'repFamily',\n",
       " 'cassetteExon',\n",
       " 'altPromoter',\n",
       " 'bleedingExon',\n",
       " 'splice score',\n",
       " 'inframe_deletion',\n",
       " 'stop_gained',\n",
       " 'frameshift_variant',\n",
       " 'splice_donor_variant',\n",
       " 'splice_acceptor_variant',\n",
       " 'splice_polypyrimidine_tract_variant',\n",
       " 'coding_sequence_variant',\n",
       " 'intron_variant',\n",
       " 'splice_region_variant',\n",
       " 'protein_altering_variant',\n",
       " '3_prime_UTR_variant',\n",
       " 'inframe_insertion',\n",
       " 'splice_donor_region_variant',\n",
       " 'splice_donor_5th_base_variant',\n",
       " 'stop_lost',\n",
       " 'start_lost',\n",
       " 'missense_variant',\n",
       " '5_prime_UTR_variant',\n",
       " 'length diff mutation',\n",
       " 'misc_RNA',\n",
       " 'nonsense_mediated_decay',\n",
       " 'promoter',\n",
       " 'protein_coding',\n",
       " 'retained_intron',\n",
       " 'Transcript',\n",
       " 'RegulatoryFeature',\n",
       " 'MotifFeature',\n",
       " 'prediction probabilities',\n",
       " 'binarized prediction']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df.columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "809011bc-2632-4ca9-b4ab-ba4d66c0fcbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Not CPG variant', 'removed', 'added'], dtype=object)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"CpG variant\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "52969e91-a595-456f-b907-fa71473c541a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'Unnamed: 0',\n",
      "'SAMPLE',\n",
      "'FILTER',\n",
      "'VAR_ID',\n",
      "'CHR',\n",
      "'POS',\n",
      "'REF',\n",
      "'ALT',\n",
      "'GENE',\n",
      "'ALLELE',\n",
      "'VARIANT_CLASS',\n",
      "'BIOTYPE',\n",
      "'TYPE',\n",
      "'FEATURE',\n",
      "'EXON',\n",
      "'HGVSc',\n",
      "'HGVSp',\n",
      "'cDNA_position',\n",
      "'CDS_position',\n",
      "'Protein_position',\n",
      "'Amino_acids',\n",
      "'Codons',\n",
      "'STRAND',\n",
      "'REFSEQ_MATCH',\n",
      "'GIVEN_REF',\n",
      "'USED_REF',\n",
      "'BAM_EDIT',\n",
      "'IMPACT',\n",
      "'CONSEQUENCE',\n",
      "'SOMATIC',\n",
      "'PHENO',\n",
      "'CLIN_SIG',\n",
      "'COSMIC_MATCH',\n",
      "'HEURISTIC',\n",
      "'PREVIOUSLY_IDENTIFIED',\n",
      "'WHITELIST',\n",
      "'ARTIFACT',\n",
      "'SIFT_DESC',\n",
      "'SIFT_SCORE',\n",
      "'POLYPHEN_DESC',\n",
      "'POLYPHEN_SCORE',\n",
      "'CADD_PHRED',\n",
      "'CADD_RAW',\n",
      "'REF_DEPTH_VIS1',\n",
      "'ALT_DEPTH_VIS1',\n",
      "'F1R2_VIS1',\n",
      "'F2R1_VIS1',\n",
      "'REF_DEPTH_VIS3',\n",
      "'ALT_DEPTH_VIS3',\n",
      "'F1R2_VIS3',\n",
      "'F2R1_VIS3',\n",
      "'VAF_VIS1',\n",
      "'VAF_VIS3',\n",
      "'VAF_ABSOLUTA',\n",
      "'VAF_RELATIVA',\n",
      "'VAF_RATIO',\n",
      "'CHIP_driver',\n",
      "'FU_yrs',\n",
      "'AER',\n",
      "'log_positive_AER',\n",
      "'log_AER',\n",
      "'N',\n",
      "'doubling_time',\n",
      "'doubling_time_wo_outliers',\n",
      "'log_ratio_VAF_years',\n",
      "'VAF_absolute_yrs',\n",
      "'growth_lm_VAF',\n",
      "'growth_lm_log_VAF',\n",
      "'fitness_exp',\n",
      "'region_normal',\n",
      "'region_alt',\n",
      "'POS.1',\n",
      "'exon',\n",
      "'Amino_acids.1',\n",
      "'IMPACT.1',\n",
      "'VARIANT_CLASS.1',\n",
      "'CpG variant',\n",
      "'PHYLO',\n",
      "'PHASTCON',\n",
      "'cancer score',\n",
      "'cancer score transcript',\n",
      "'cys-rich domain',\n",
      "'DSBH',\n",
      "'disordered',\n",
      "'phosphoserine',\n",
      "'phosphothreonine',\n",
      "'polar residues',\n",
      "'lys-gly crosslink',\n",
      "'Assymetric dimethylargine',\n",
      "'chain',\n",
      "'interaction with DNA',\n",
      "'SPI1_score',\n",
      "'RAD21_score',\n",
      "'GATA2_score',\n",
      "'GATA3_score',\n",
      "'YY1_score',\n",
      "'EP300_score',\n",
      "'BATF_score',\n",
      "'CTCF_score',\n",
      "'EED_score',\n",
      "'JUNB_score',\n",
      "'POLR2A_score',\n",
      "'POU2F2_score',\n",
      "'IRF4_score',\n",
      "'RCOR1_score',\n",
      "'SMARCA4_score',\n",
      "'Dnase cluster',\n",
      "'GC%',\n",
      "'CPG',\n",
      "'UKB',\n",
      "'H3K4Me1',\n",
      "'H3K4Me3',\n",
      "'H3K27ac',\n",
      "'cCREs score',\n",
      "'cCREs label',\n",
      "'geneHancer score',\n",
      "'geneHancer elementType',\n",
      "'swScore',\n",
      "'repFamily',\n",
      "'cassetteExon',\n",
      "'altPromoter',\n",
      "'bleedingExon',\n",
      "'splice score',\n",
      "'inframe_deletion',\n",
      "'stop_gained',\n",
      "'frameshift_variant',\n",
      "'splice_donor_variant',\n",
      "'splice_acceptor_variant',\n",
      "'splice_polypyrimidine_tract_variant',\n",
      "'coding_sequence_variant',\n",
      "'intron_variant',\n",
      "'splice_region_variant',\n",
      "'protein_altering_variant',\n",
      "'3_prime_UTR_variant',\n",
      "'inframe_insertion',\n",
      "'splice_donor_region_variant',\n",
      "'splice_donor_5th_base_variant',\n",
      "'stop_lost',\n",
      "'start_lost',\n",
      "'missense_variant',\n",
      "'5_prime_UTR_variant',\n",
      "'length diff mutation',\n",
      "'misc_RNA',\n",
      "'nonsense_mediated_decay',\n",
      "'promoter',\n",
      "'protein_coding',\n",
      "'retained_intron',\n",
      "'Transcript',\n",
      "'RegulatoryFeature',\n",
      "'MotifFeature',\n",
      "'prediction probabilities',\n",
      "'binarized prediction',\n",
      "                         shapiro_p_chip shapiro_p_nonchip levene_p_value  \\\n",
      "Unnamed: 0                          0.0          0.000003       0.901818   \n",
      "POS                                 0.0               0.0            0.0   \n",
      "REFSEQ_MATCH                        NaN               NaN            NaN   \n",
      "exon                                NaN               NaN            NaN   \n",
      "Amino_acids.1                       NaN               NaN            NaN   \n",
      "...                                 ...               ...            ...   \n",
      "Transcript                          1.0               1.0            NaN   \n",
      "RegulatoryFeature                   1.0               1.0            NaN   \n",
      "MotifFeature                        1.0               1.0            NaN   \n",
      "prediction probabilities            0.0               0.0       0.000064   \n",
      "binarized prediction                0.0               0.0       0.004259   \n",
      "\n",
      "                            test_used test_p_value  \n",
      "Unnamed: 0                mannwhitney          0.0  \n",
      "POS                       mannwhitney          0.0  \n",
      "REFSEQ_MATCH              mannwhitney          NaN  \n",
      "exon                      mannwhitney          NaN  \n",
      "Amino_acids.1             mannwhitney          NaN  \n",
      "...                               ...          ...  \n",
      "Transcript                mannwhitney          1.0  \n",
      "RegulatoryFeature         mannwhitney          1.0  \n",
      "MotifFeature              mannwhitney          1.0  \n",
      "prediction probabilities  mannwhitney          0.0  \n",
      "binarized prediction      mannwhitney          0.0  \n",
      "\n",
      "[81 rows x 5 columns]\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "\n",
    "ignore = ['SAMPLE','FILTER','VAR_ID','CHR','REF','ALT','GENE', 'ALLELE','VARIANT_CLASS','BIOTYPE','TYPE','HGVSc','HGVSp','EXON','cDNA_position',\n",
    " 'CDS_position','Amino_acids','Codons','Protein_position', 'GIVEN_REF','USED_REF','BAM_EDIT','IMPACT','CONSEQUENCE','SOMATIC',\n",
    " 'PHENO','CLIN_SIG','COSMIC_MATCH','HEURISTIC','PREVIOUSLY_IDENTIFIED','WHITELIST','ARTIFACT','SIFT_DESC','SIFT_SCORE','POLYPHEN_DESC',\n",
    " 'POLYPHEN_SCORE','CADD_PHRED','CADD_RAW','REF_DEPTH_VIS1','ALT_DEPTH_VIS1','F1R2_VIS1','F2R1_VIS1','REF_DEPTH_VIS3',\n",
    " 'ALT_DEPTH_VIS3','F1R2_VIS3','F2R1_VIS3','VAF_VIS1','VAF_VIS3','VAF_ABSOLUTA','VAF_RELATIVA','VAF_RATIO','CHIP_driver',\n",
    " 'FU_yrs','AER','log_positive_AER','log_AER','N','doubling_time','doubling_time_wo_outliers','log_ratio_VAF_years','VAF_absolute_yrs','growth_lm_VAF',\n",
    " 'growth_lm_log_VAF','fitness_exp','region_normal','region_alt','POS.1',\"FEATURE\", \"STRAND\", \"IMPACT.1\", 'VARIANT_CLASS.1',]\n",
    "df[\"CpG variant\"] = df[\"CpG variant\"].map({'Not CPG variant':0, 'removed':-1, 'added':1})\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "tf = pd.DataFrame()\n",
    "for item in df.columns:\n",
    "    print(f\"'{item}',\")\n",
    "    if item in ignore:\n",
    "        continue\n",
    "    positive = df[df[\"CHIP_driver\"] == \"YES\"][item]\n",
    "    negative = df[df[\"CHIP_driver\"] == \"NO\"][item]\n",
    "    \n",
    "    # shapiro wilk drivers\n",
    "    _, pos_shapiro_p = stats.shapiro(positive)\n",
    "    _, neg_shapiro_p = stats.shapiro(negative)\n",
    "    \n",
    "    # shapiro wilk non drivers\n",
    "    _, levene_p = stats.levene(positive, negative)\n",
    "    \n",
    "    # if normal distribution for both use t test, else use a mannwhitney test\n",
    "    if pos_shapiro_p > 0.05 and neg_shapiro_p > 0.05 and levene_p > 0.05:\n",
    "        _, ttest_p = stats.ttest_ind(positive, negative, equal_var=True)\n",
    "        test_type = \"t_test\"\n",
    "        test_p = ttest_p\n",
    "    else:\n",
    "        _, mannwhitney_p = stats.mannwhitneyu(positive, negative, alternative='two-sided')\n",
    "        test_type = \"mannwhitney\"\n",
    "        test_p = mannwhitney_p\n",
    "        tf[item] = [pos_shapiro_p, neg_shapiro_p, levene_p, test_type, test_p]\n",
    "\n",
    "tf.index = [\"shapiro_p_chip\", \"shapiro_p_nonchip\", \"levene_p_value\", \"test_used\", \"test_p_value\"]\n",
    "\n",
    "tf = tf.transpose()\n",
    "print(tf)\n",
    "tf.to_excel(\"./statistics.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "37bfe604-94f7-443d-a16e-92644193bb4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf.to_excel(\"./importances_tf.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "679112e7-afda-4780-84af-24b867108eb2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "               shapiro_p_chip  shapiro_p_nonchip  mannwhitney_p_value\n",
      "USF2_score       1.000000e+00       1.029730e-31         5.939461e-08\n",
      "RNF2_score       1.000000e+00       1.431137e-31         2.613368e-08\n",
      "REST_score       1.000000e+00       1.029730e-31         5.939461e-08\n",
      "RAD21_score      1.000000e+00       3.865148e-34         5.928925e-03\n",
      "BRD4_score       1.000000e+00       3.523085e-35         2.206701e-01\n",
      "...                       ...                ...                  ...\n",
      "RUNX3_score      1.000000e+00       7.130173e-35         8.267866e-02\n",
      "BHLHE40_score    1.000000e+00       3.865148e-34         5.928925e-03\n",
      "SIN3A_score      1.000000e+00       1.431137e-31         2.613368e-08\n",
      "H3K4Me1          2.704870e-21       1.718032e-24         3.005362e-01\n",
      "H3K4Me3          2.093050e-17       7.648820e-24         2.490026e-01\n",
      "\n",
      "[61 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# Initialize the DataFrame to store the results\n",
    "tf = pd.DataFrame()\n",
    "\n",
    "# Loop through each column in tf_columns\n",
    "for item in tf_columns:\n",
    "    positive = df[df[\"binarized prediction\"] == 1][item]\n",
    "    negative = df[df[\"binarized prediction\"] == 0][item]\n",
    "    \n",
    "    # Perform Shapiro-Wilk test\n",
    "    _, pos_shapiro_p = stats.shapiro(positive)\n",
    "    _, neg_shapiro_p = stats.shapiro(negative)\n",
    "    \n",
    "    # Perform Mann-Whitney U test\n",
    "    _, mannwhitney_p = stats.mannwhitneyu(positive, negative, alternative='two-sided')\n",
    "    \n",
    "    # Store the results in the DataFrame\n",
    "    tf[item] = [pos_shapiro_p, neg_shapiro_p, mannwhitney_p]\n",
    "\n",
    "# Rename the rows for clarity\n",
    "tf.index = [\"shapiro_p_chip\", \"shapiro_p_nonchip\", \"mannwhitney_p_value\"]\n",
    "\n",
    "# Transpose the DataFrame to have items as columns and tests as rows\n",
    "tf = tf.transpose()\n",
    "\n",
    "# Display the DataFrame\n",
    "print(tf)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "067b831c-0098-408b-9f18-a8856dd26f91",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf.to_excel(\"./importances_tf_model_classifications.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94a4d59d-9086-4d3a-82c2-acfac47a3c82",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
